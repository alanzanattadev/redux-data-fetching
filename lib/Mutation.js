// @flow

export default class Mutation {
  mutationQL: string;
  variables: { [variableName: string]: mixed };
  onError: () => void;
  onCompleted: () => void;
  operationName: ?string;

  constructor(
    {
      mutationQL,
      variables = {},
      onError = () => {},
      onCompleted = () => {},
      operationName,
    }: {
      mutationQL: string,
      variables?: { [variableName: string]: mixed },
      onError?: () => void,
      onCompleted?: () => void,
      operationName?: string,
    } = {},
  ) {
    if (mutationQL == null) {
      throw new Error(
        "mutationQL has to be defined in the params of new Mutation: new Mutation({mutationQL: `mutation ...`})",
      );
    }
    this.mutationQL = mutationQL;
    this.variables = variables;
    this.onError = onError;
    this.onCompleted = onCompleted;
    this.operationName = operationName;
  }

  setOperationName(name: string) {
    return new Mutation({
      mutationQL: this.mutationQL,
      variables: this.variables,
      onError: this.onError,
      onCompleted: this.onCompleted,
      operationName: name,
    });
  }
}
