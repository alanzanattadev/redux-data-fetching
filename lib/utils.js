// @flow

import { Stack, List } from "immutable";
import {
  TypeInfo,
  GraphQLSchema,
  Kind,
  GraphQLList,
  GraphQLObjectType,
  getNamedType,
  visit,
  visitWithTypeInfo,
  parse,
} from "graphql";
import type { ASTNode, FieldNode } from "graphql";
import { DataReducerRecord } from "./reducer";
import { isEntity } from "./graphqlTypesConverters";

export function hashString(str: string): number {
  let hash = 0;
  if (str.length == 0) return hash;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash; // Convert to 32bit integer
  }
  return hash;
}

export class TypeInfoWithValuesComparator extends TypeInfo {
  _value1Stack: Stack<mixed>;
  _value2Stack: Stack<mixed>;
  _reducer1: DataReducerRecord;
  _reducer2: DataReducerRecord;
  _queryHash1: string | number;
  _queryHash2: string | number;
  _shouldBreak: boolean;

  constructor({
    schema,
    reducer1,
    reducer2,
    queryHash1,
    queryHash2,
  }: {
    schema: GraphQLSchema,
    reducer1: DataReducerRecord,
    reducer2: DataReducerRecord,
    queryHash1: string | number,
    queryHash2: string | number,
  }) {
    super(schema);
    this._value1Stack = Stack();
    this._value2Stack = Stack();
    this._reducer1 = reducer1;
    this._reducer2 = reducer2;
    this._queryHash1 = queryHash1;
    this._queryHash2 = queryHash2;
    this._shouldBreak = false;
  }

  _getQueryValue(
    node: FieldNode,
    reducer: DataReducerRecord,
    queryHash: string | number,
  ) {
    if (reducer == null || queryHash == null)
      throw new Error(
        "Trying to determine whether selection of data has to be relaunched, but encounterd an error. Missing dependency on reducer or queryHash",
      );
    const queryResult = reducer.getIn([
      "queries",
      queryHash,
      "results",
      "byQuery",
      node.alias != null ? node.alias.value : node.name.value,
    ]);
    const currentType = this.getType();
    if (queryResult == null) {
      if (currentType instanceof GraphQLList) return [];
      else return null;
    }
    if (currentType instanceof GraphQLList) {
      const typeName = getNamedType(currentType.ofType).name;
      return queryResult.map(id =>
        reducer.getIn(["entities", typeName, id.toString()]),
      );
    } else if (queryResult != null) {
      const namedType = getNamedType(currentType);
      if (namedType != undefined) {
        return reducer.getIn([
          "entities",
          namedType.name,
          queryResult.toString(),
        ]);
      } else {
        console.error("CurrentType:", currentType);
        throw new Error(
          "Weird behavior has been encountered with an unnamed type as current type",
        );
      }
    } else {
      return null;
    }
  }

  _getEntityValue(
    node: FieldNode,
    reducer: DataReducerRecord,
    lastValue: mixed,
  ) {
    if (
      node.selectionSet != null &&
      typeof lastValue === "object" &&
      lastValue != null
    ) {
      const fieldName = node.alias ? node.alias.value : node.name.value;
      if (Array.isArray(lastValue) || List.isList(lastValue))
        // $FlowFixMe
        return lastValue.map(v => this._getEntityValue(node, reducer, v));
      const fieldValue = lastValue[fieldName];
      const namedType = getNamedType(this.getType());
      if (namedType && isEntity(namedType) === false) {
        return fieldValue;
      }
      if (
        this.getType() instanceof GraphQLList &&
        fieldValue != null &&
        (Array.isArray(fieldValue) || List.isList(fieldValue))
      ) {
        if (namedType) {
          const entityName = namedType.name;
          // $FlowFixMe
          return fieldValue.map(id =>
            reducer.getIn(["entities", entityName, id.toString()]),
          );
        } else {
          console.error(
            "Weird state encountered trying to get the named type of",
            this.getType(),
            "at node",
            node,
            "knowing that ",
            fieldValue,
            "is an Array or a Immutable.List",
          );
          throw new Error("Aborting.");
        }
      } else if (fieldValue != null) {
        if (namedType) {
          const entityName = namedType.name;
          // $FlowFixMe
          return reducer.getIn(["entities", entityName, fieldValue.toString()]);
        } else {
          console.error(
            "Weird state encountered trying to get the named type of",
            this.getType(),
            "at node",
            node,
            "knowing that ",
            fieldValue,
            "is an Object or a Immutable.Record",
          );
          throw new Error("Aborting.");
        }
      } else {
        return null;
      }
    } else {
      return lastValue;
    }
  }

  _haveChanged(value1: mixed, value2: mixed): boolean {
    if (typeof value1 !== typeof value2) return true;
    if (Array.isArray(value1) || List.isList(value1)) {
      if (Array.isArray(value2) === true || List.isList(value2) === true) {
        // $FlowFixMe
        return value1.some((v1: mixed, i: number) => {
          return this._haveChanged(
            v1,
            // $FlowFixMe
            Array.isArray(value2) ? value2[i] : value2.get(i),
          );
        });
      } else return true;
    } else if (typeof value1 === "object" && value1 !== null) {
      if (value2 === null) return true;
      return value1 !== value2;
    } else {
      return value1 !== value2;
    }
  }

  enter(node: ASTNode) {
    super.enter(node);
    if (node.kind === Kind.FIELD) {
      let value1;
      let value2;
      if (this.getParentType() === this._schema.getQueryType()) {
        value1 = this._getQueryValue(node, this._reducer1, this._queryHash1);
        this._value1Stack = this._value1Stack.unshift(value1);
        value2 = this._getQueryValue(node, this._reducer2, this._queryHash2);
        this._value2Stack = this._value2Stack.push(value2);
      } else {
        value1 = this._getEntityValue(
          node,
          this._reducer1,
          this._value1Stack.peek(),
        );
        this._value1Stack = this._value1Stack.push(value1);
        value2 = this._getEntityValue(
          node,
          this._reducer2,
          this._value2Stack.peek(),
        );
        this._value2Stack = this._value2Stack.push(value2);
      }
      if (this._shouldBreak === false) {
        this._shouldBreak = this._haveChanged(value1, value2);
      }
    }
  }

  leave(node: ASTNode) {
    if (node.kind === Kind.FIELD) {
      this._value1Stack = this._value1Stack.pop();
      this._value2Stack = this._value2Stack.pop();
    }
    super.leave(node);
  }

  getValue1(): mixed {
    return this._value1Stack.peek();
  }

  getValue2(): mixed {
    return this._value2Stack.peek();
  }

  shouldBreak(): boolean {
    return this._shouldBreak;
  }
}

export function selectedDataHaveChanged({
  query,
  schema,
  queryHash,
  reducer1,
  reducer2,
}: {
  query: string,
  schema: GraphQLSchema,
  queryHash: string | number,
  reducer1: DataReducerRecord,
  reducer2: DataReducerRecord,
}): boolean {
  const typeInfo = new TypeInfoWithValuesComparator({
    schema,
    reducer1,
    reducer2,
    queryHash1: queryHash,
    queryHash2: queryHash,
  });
  let hasChanged = false;
  visit(
    parse(query),
    visitWithTypeInfo(typeInfo, {
      [Kind.FIELD]: {
        enter(node) {
          hasChanged = typeInfo.shouldBreak();
        },
      },
    }),
  );
  return hasChanged;
}
