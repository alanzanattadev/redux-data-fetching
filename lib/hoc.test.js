import React from "react";
import { fromJS, Map, List, Record } from "immutable";
import configureGraphQLConnecter from "./hoc";
import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString,
  GraphQLList,
} from "graphql";
import { mount } from "enzyme";
import { compose, lifecycle } from "recompose";
import {
  DataReducerRecord,
  QueryRecord,
  ResultsRecord,
  QUERY_PROGRESS_PENDING,
} from "./reducer";
import { hashString } from "./utils";
import configureActions from "./actions";
import Mutation from "./Mutation";

const User = new GraphQLObjectType({
  name: "User",
  fields: () => ({
    id: { type: GraphQLString },
    name: { type: GraphQLString },
    description: { type: GraphQLString },
    friends: { type: new GraphQLList(User) },
  }),
});

describe("HOC", () => {
  const actions = configureActions();
  const { GraphQLConnecter, DataHandlers } = configureGraphQLConnecter({
    actions,
    typesSchema: new GraphQLSchema({
      query: new GraphQLObjectType({
        name: "RootQuery",
        fields: {
          user: {
            type: User,
            resolve: () => {},
            args: { id: { type: GraphQLString } },
          },
          users: {
            type: new GraphQLList(User),
            resolve: (p, args, context) =>
              context.db && context.queryHash ? [] : null,
          },
        },
      }),
    }),
    selectorSchema: new GraphQLSchema({
      query: new GraphQLObjectType({
        name: "RootQuery",
        fields: {
          user: {
            type: User,
            resolve: () => {},
            args: { id: { type: GraphQLString } },
          },
          users: {
            type: new GraphQLList(User),
            resolve: (p, args, context) =>
              context.db && context.queryHash ? [] : null,
          },
        },
      }),
    }),
    recordTypes: {
      User: Record({
        id: null,
        name: null,
        description: null,
        friends: null,
      }),
    },
  });

  describe("DataHandlers", () => {
    it("passes others props to the wrapped components", () => {
      function Comp({ testProp }) {
        expect(testProp).toBeDefined();
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({}),
      })(Comp);

      mount(
        <ConnectedComp
          testProp={true}
          data={new DataReducerRecord()}
          dispatch={() => {}}
        />,
      );
    });

    it("throws if dispatch is not passed as props", () => {
      function Comp({ testProp }) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({}),
      })(Comp);

      expect(() =>
        mount(<ConnectedComp data={{}} />),
      ).toThrowErrorMatchingSnapshot();
    });

    it("throws if reducer is not passed as props", () => {
      function Comp({ testProp }) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({}),
      })(Comp);

      expect(() =>
        mount(<ConnectedComp dispatch={() => {}} />),
      ).toThrowErrorMatchingSnapshot();
    });

    it("passes the handlers as props", () => {
      function Comp({ onClick }) {
        expect(onClick).toBeDefined();
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          onClick: () => new Mutation({ mutationQL: "mutation {}" }),
        }),
      })(Comp);

      mount(
        <ConnectedComp data={new DataReducerRecord()} dispatch={() => {}} />,
      );
    });

    it("dispatches mutations when a handler is called", () => {
      function Comp({}) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          onClick: () => new Mutation({ mutationQL: "mutation {}" }),
        }),
      })(Comp);
      const spy = jest.fn();

      const subject = mount(
        <ConnectedComp
          __uniqueID={2198392}
          data={new DataReducerRecord()}
          dispatch={spy}
        />,
      );

      subject.find(Comp).at(0).prop("onClick")();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls).toMatchSnapshot();
    });
    it("doesn't dispatch the mutation if returned mutation is null", () => {
      function Comp({}) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          onClick: () => null,
        }),
      })(Comp);
      const spy = jest.fn();

      const subject = mount(
        <ConnectedComp data={new DataReducerRecord()} dispatch={spy} />,
      );

      subject.find(Comp).at(0).prop("onClick")();

      expect(spy).not.toHaveBeenCalled();
    });

    it("throws if a handler isn't a function", () => {
      function Comp({ testProp }) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          onClick: 2,
        }),
      })(Comp);

      expect(() =>
        mount(
          <ConnectedComp dispatch={() => {}} data={new DataReducerRecord()} />,
        ),
      ).toThrowErrorMatchingSnapshot();
    });

    it("passes the query progress as props", () => {
      function Comp({ createUserQueryProgress, onClickQueryProgress }) {
        expect(createUserQueryProgress).toMatchSnapshot();
        expect(onClickQueryProgress).toMatchSnapshot();
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          onClick: () => new Mutation({ mutationQL: "mutation {}" }),
          createUser: () => new Mutation({ mutationQL: "mutation {}" }),
        }),
      })(Comp);

      mount(
        <ConnectedComp
          __uniqueID={2198392}
          data={
            new DataReducerRecord({
              queries: Map({
                "1385894182": new QueryRecord({
                  progress: QUERY_PROGRESS_PENDING,
                }),
                "436150438": new QueryRecord({}),
              }),
            })
          }
          dispatch={() => {}}
        />,
      );
    });

    it("passes the bustQueryCache as props", () => {
      function Comp({ bustQueryCache }) {
        expect(bustQueryCache).toBeDefined();
        expect(typeof bustQueryCache).toBe("function");
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          createUser: () => new Mutation({ mutationQL: "mutation {}" }),
        }),
      })(Comp);

      mount(
        <ConnectedComp data={new DataReducerRecord()} dispatch={() => {}} />,
      );
    });

    it("dispatches an action when bustQueryCache is called", () => {
      function Comp({}) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({}),
      })(Comp);
      const spy = jest.fn();

      const subject = mount(
        <ConnectedComp
          __uniqueID={2198392}
          data={new DataReducerRecord()}
          dispatch={spy}
        />,
      );

      subject.find(Comp).at(0).prop("bustQueryCache")();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls).toMatchSnapshot();
    });

    it("busts cache on unmount", () => {
      function Comp({}) {
        return null;
      }
      const ConnectedComp = DataHandlers({
        mapMutationsToProps: () => ({
          onClick: () => new Mutation({ mutationQL: "" }),
          onRemove: () => new Mutation({ mutationQL: "" }),
        }),
      })(Comp);
      const spy = jest.fn();

      const subject = mount(
        <ConnectedComp
          __uniqueID={2198392}
          data={
            new DataReducerRecord({
              queries: Map({
                "1385894182": new QueryRecord({}),
                "436150438": new QueryRecord({}),
              }),
            })
          }
          dispatch={spy}
        />,
      );

      subject.unmount();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls).toMatchSnapshot();
    });
  });

  describe("GraphQLConnecter", () => {
    it("should fetch data on mount", () => {
      let MyComponent = ({ name, description }) =>
        <div>
          <h1>
            {name}
          </h1>
          <p>
            {description}
          </p>
        </div>;
      let spy = jest.fn();
      let MyFetcherComponent = GraphQLConnecter(
        props => `{user(id: "${props.id}") {name, description}}`,
        (cache, props) => cache.getIn(["users", props.id]).toJS(),
      )(MyComponent);
      let subject = mount(
        <MyFetcherComponent
          dispatch={spy}
          data={fromJS({
            users: { "1": { name: "Jon", description: "Good" } },
          })}
          id={"1"}
        />,
      );

      expect(spy).toBeCalledWith({
        type: "GRAPHQL_FETCH",
        graphql: true,
        payload: '{user(id: "1") {name, description}}',
      });
    });

    it("should display with computed props", () => {
      let MyComponent = ({ name, description }) =>
        <div>
          <h1>
            {name}
          </h1>
          <p>
            {description}
          </p>
        </div>;
      let MyFetcherComponent = GraphQLConnecter(
        props => `{user(id: "${props.id}") {name, description}}`,
        (cache, props) => cache.getIn(["users", props.id]).toJS(),
      )(MyComponent);
      let subject = mount(
        <MyFetcherComponent
          dispatch={() => {}}
          data={fromJS({
            users: { "1": { name: "Jon", description: "Good" } },
          })}
          id={"1"}
        />,
      );

      expect(subject.find(MyComponent).at(0).prop("name")).toBe("Jon");
      expect(subject.find(MyComponent).at(0).prop("description")).toBe("Good");
    });

    it("should refetch data when needs change", () => {
      let MyComponent = ({ name, description }) =>
        <div>
          <h1>
            {name}
          </h1>
          <p>
            {description}
          </p>
        </div>;
      let spy = jest.fn();
      let data = fromJS({
        users: {
          "1": { name: "Jon", description: "Good" },
          "2": { name: "Doe", description: "Bad" },
        },
      });
      let MyFetcherComponent = GraphQLConnecter(
        props => `{user(id: "${props.id}") {name, description}}`,
        (cache, props) => cache.getIn(["users", props.id]),
      )(MyComponent);
      let subject = mount(
        <MyFetcherComponent dispatch={() => {}} data={data} id={"1"} />,
      );

      subject.setProps({ dispatch: spy, data: data, id: "2" });

      expect(spy).toBeCalledWith({
        type: "GRAPHQL_FETCH",
        graphql: true,
        payload: '{user(id: "2") {name, description}}',
      });
    });

    it("should refetch data on call of refetch props", () => {
      let MyComponent = ({ name, description }) =>
        <div>
          <h1>
            {name}
          </h1>
          <p>
            {description}
          </p>
        </div>;
      let spy = jest.fn();
      let data = fromJS({
        users: {
          "1": { name: "Jon", description: "Good" },
          "2": { name: "Doe", description: "Bad" },
        },
      });
      let MyFetcherComponent = GraphQLConnecter(
        props => `{user(id: "1") {name, description}}`,
        (cache, props) => cache.getIn(["users", props.id]),
      )(MyComponent);
      let subject = mount(
        <MyFetcherComponent dispatch={() => {}} data={data} id={"1"} />,
      );

      subject.setProps({ dispatch: spy, data: data });
      subject.find(MyComponent).at(0).prop("refetch")();

      expect(spy).toBeCalledWith({
        type: "GRAPHQL_FETCH",
        graphql: true,
        payload: '{user(id: "1") {name, description}}',
      });
    });

    it("selects data in the cache", done => {
      const request = `{ users { id, name, friends { id, name, friends { id, name } } } }`;
      const hash = hashString(request);
      const User = compose(
        GraphQLConnecter(() => request),
        lifecycle({
          componentDidUpdate() {
            if (List.isList(this.props.users)) {
              expect(this.props).toMatchSnapshot();
              done();
            }
          },
        }),
      )(({ users }) => <div />);
      const dataReducer = new DataReducerRecord({
        entities: Map({
          User: Map({
            1: {
              id: 1,
              name: "Alan",
              friends: [2, 3],
            },
            2: {
              id: 2,
              name: "Antoine",
              friends: [1],
            },
            3: {
              id: 3,
              name: "Sophie",
              friends: null,
            },
          }),
        }),
        queries: Map({
          [hash]: new QueryRecord({
            results: new ResultsRecord({
              byQuery: Map({
                users: [1, 2],
              }),
            }),
          }),
        }),
      });
      const subject = mount(<User data={dataReducer} dispatch={() => {}} />);
    });
  });
});
