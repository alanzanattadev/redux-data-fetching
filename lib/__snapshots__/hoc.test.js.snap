// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC DataHandlers busts cache on unmount 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "queryID": 1385894182,
      },
      "type": "QUERY_CACHE_BUSTED",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "queryID": 436150438,
      },
      "type": "QUERY_CACHE_BUSTED",
    },
  ],
]
`;

exports[`HOC DataHandlers dispatches an action when bustQueryCache is called 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "queryID": 1464482893,
      },
      "type": "QUERY_CACHE_BUSTED",
    },
  ],
]
`;

exports[`HOC DataHandlers dispatches mutations when a handler is called 1`] = `
Array [
  Array [
    Object {
      "graphql": true,
      "payload": Object {
        "mutation": Mutation {
          "mutationQL": "mutation {}",
          "onCompleted": [Function],
          "onError": [Function],
          "operationName": undefined,
          "variables": Object {},
        },
        "queryID": 1385894182,
      },
      "type": "GRAPHQL_MUTATION",
    },
  ],
]
`;

exports[`HOC DataHandlers dispatches mutations when a handler is called 2`] = `
Array [
  Array [
    "a",
    "b",
  ],
]
`;

exports[`HOC DataHandlers passes the query progress as props 1`] = `"not started"`;

exports[`HOC DataHandlers passes the query progress as props 2`] = `"pending"`;

exports[`HOC DataHandlers throws if a handler isn't a function 1`] = `"You must pass a function as handler of mapMutationsToProps, handler onClick isn't a function"`;

exports[`HOC DataHandlers throws if dispatch is not passed as props 1`] = `"DataHandlers must get the dispatch function as props"`;

exports[`HOC DataHandlers throws if reducer is not passed as props 1`] = `"DataHandlers must get the cache reducer as a props named 'data'"`;

exports[`HOC GraphQLConnecter selects data in the cache 1`] = `
Object {
  "data": Object {
    "entities": Object {
      "User": Object {
        "1": Object {
          "friends": Array [
            2,
            3,
          ],
          "id": 1,
          "name": "Alan",
        },
        "2": Object {
          "friends": Array [
            1,
          ],
          "id": 2,
          "name": "Antoine",
        },
        "3": Object {
          "friends": null,
          "id": 3,
          "name": "Sophie",
        },
      },
    },
    "queries": Object {
      "1002205036": Object {
        "progress": "not started",
        "results": Object {
          "byEntity": Object {},
          "byQuery": Object {
            "users": Array [
              1,
              2,
            ],
          },
        },
      },
    },
  },
  "dispatch": [Function],
  "queryProgress": "not started",
  "refetch": [Function],
  "users": Immutable.List [
],
}
`;
